File: java/108.convert-sorted-array-to-binary-search-tree.java
/*
 * @lc app=leetcode id=108 lang=java
 *
 * [108] Convert Sorted Array to Binary Search Tree
 */
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode sortedArrayToBST(int[] nums) {
        if (nums == null || nums.length == 0) return null;
        return builder(nums, 0, nums.length-1);
    }

    private TreeNode builder(int[] nums, int l, int h) {
        int m = (l + h) / 2;
        TreeNode node = new TreeNode(nums[m]);
        if (l == h) return node;
        if (h - l == 1) {
            TreeNode tmp = new TreeNode(nums[h]);
            tmp.left = node;
            return tmp;
        }
        node.left = builder(nums, l, m-1);
        node.right = builder(nums, m+1, h);
        return node;
    }
}

File: go/108.convert-sorted-array-to-binary-search-tree.go
/*
 * @lc app=leetcode id=108 lang=golang
 *
 * [108] Convert Sorted Array to Binary Search Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func sortedArrayToBST(nums []int) *TreeNode {
	if len(nums) == 0 {
		return nil
	}
    if len(nums) == 1 {
		return &TreeNode{nums[0], nil, nil}
	}
	if len(nums) == 2 {
		return &TreeNode{nums[1], &TreeNode{nums[0], nil, nil}, nil}
	}
	return &TreeNode{nums[len(nums)/2], 
		sortedArrayToBST(nums[:len(nums)/2]), 
		sortedArrayToBST(nums[len(nums)/2+1:])}

}
// @lc code=end

